snapshot-cassandra/src/main/scala/com/evolutiongaming/kafka/journal/snapshot/cassandra/SnapshotCassandra.scala

```diff
@@ -0,0 +1,172 @@
+          case s if s.size < numberOfSnapshots => insert(key, s, snapshot)
+          // all rows are taken, we have to update one of them
+          case s => replace(key, s, snapshot)
+        }
+      }
+
+      def insert(
+        key: Key,
+        savedSnapshots: Map[BufferNr, (SeqNr, Instant)],
+        snapshot: SnapshotRecord[EventualPayloadAndType]
+      ): F[Unit] = {
+        val allBufferNrs = BufferNr.listOf(numberOfSnapshots)
+        val takenBufferNrs = savedSnapshots.keySet
+        val freeBufferNr = allBufferNrs.find(bufferNr => !takenBufferNrs.contains(bufferNr))
+        MonadThrow[F].fromOption(freeBufferNr, SnapshotStoreError("Could not find a free key")).flatMap { bufferNr =>
+          val wasApplied = statements.insertRecord(key, bufferNr, snapshot)
+          wasApplied.flatMap { wasApplied =>
+            // TODO: consider adding circuit breaker here
+            if (wasApplied) ().pure[F] else save(key, snapshot)
+          }
+        }
+      }
+
+      def replace(
+        key: Key,
+        savedSnapshots: Map[BufferNr, (SeqNr, Instant)],
+        insertSnapshot: SnapshotRecord[EventualPayloadAndType]
+      ): F[Unit] = {
+        val oldestSnapshot = savedSnapshots.toList.minByOption { case (_, (seqNr, timestamp)) => (seqNr, timestamp) }
+
+        MonadThrow[F].fromOption(oldestSnapshot, SnapshotStoreError("Could not find an oldest snapshot")).flatMap {
+          oldestSnapshot =>
+            val (bufferNr, (deleteSnapshot, _)) = oldestSnapshot
+            val wasApplied = statements.updateRecord(key, bufferNr, insertSnapshot, deleteSnapshot)
+            wasApplied.flatMap { wasApplied =>
+              // TODO: consider adding circuit breaker here
+              if (wasApplied) ().pure[F] else save(key, insertSnapshot)
+            }
+        }
+      }
+
+      def update(
+        key: Key,
+        savedSnapshots: Map[BufferNr, (SeqNr, Instant)],
+        insertSnapshot: SnapshotRecord[EventualPayloadAndType]
+      ): F[Unit] = {
+        val sortedSnapshots = savedSnapshots.toList.sortBy { case (_, (seqNr, timestamp)) => (seqNr, timestamp) }
+
+        val olderSnapshot =
+          sortedSnapshots.find { case (_, (seqNr, _)) => seqNr == insertSnapshot.snapshot.seqNr }
+        MonadThrow[F].fromOption(olderSnapshot, SnapshotStoreError("Could not find snapshot with seqNr")).flatMap {
+          olderSnapshot =>
+            val (bufferNr, (deleteSnapshot, _)) = olderSnapshot
+            val wasApplied = statements.updateRecord(key, bufferNr, insertSnapshot, deleteSnapshot)
+            wasApplied.flatMap { wasApplied =>
+              // TODO: consider adding circuit breaker here
+              if (wasApplied) ().pure[F] else save(key, insertSnapshot)
+            }
+        }
+      }
+
+      def load(key: Key, criteria: SnapshotSelectionCriteria): F[Option[SnapshotRecord[EventualPayloadAndType]]] =
+        for {
+          savedSnapshots <- statements.selectMetadata(key)
```

Comment:
Would it be possible to select snapshot without querying metadata first? I did not check all details, but criteria looks like something that can be expressed in terms of CQL

Reply:
I was trying to avoid `ALLOW FILTERING` here.

Saying that, as we do have primary columns in `WHERE` clause then CQL might, indeed, work here. :thinking: